### PR TITLE
ref(cardinality): Split Redis module into multiple smaller modules

### PR DESCRIPTION
**Pull request by Dav1dde**
Splits the `relay_cardiniality::redis` module into multiple smaller modules, removing functionality from the module which was not necessarily Redis specific.

All code changes are marked with a github comment.

#skip-changelog

### CODE REVIEW

File: relay-cardinality/src/cache.rs [hunk 1]
```diff
@@ -0,0 +1,251 @@
...
+                })
+                .map_or(CacheOutcome::Unknown, |s| s.check(hash, *limit)),
+        }
+    }
+}
+
+/// Cache update handle.
+///
+/// Holds a cache write lock, the lock is released on drop.
+pub struct CacheUpdate<'a>(CacheUpdateInner<'a>);
```
Comment (Dav1dde) [inline relay-cardinality/src/cache.rs:133]: Wrapper struct is new, `CacheUpdateInner` can't be public without making `Inner` public, which is not what we want.

File: relay-cardinality/src/cache.rs [hunk 2]
```diff
@@ -0,0 +1,251 @@
...
+            item_scope,
+        };
+        CacheUpdate::new(inner, key)
+    }
+}
+
+/// Cache read handle.
+///
+/// Holds a cache read lock, the lock is released on drop.
+pub struct CacheRead<'a>(CacheReadInner<'a>);
```
Comment (Dav1dde) [inline relay-cardinality/src/cache.rs:81]: Wrapper struct is new, `CacheReadInner` can't be public without making `Inner` public, which is not what we want.

File: relay-cardinality/src/cache.rs [hunk 3]
```diff
@@ -0,0 +1,251 @@
...
+
+/// Internal state for [`CacheRead`].
+enum CacheReadInner<'a> {
+    Noop,
```
Comment (Dav1dde) [inline relay-cardinality/src/cache.rs:85]: Renamed from `Empty`

File: relay-cardinality/src/cache.rs [hunk 4]
```diff
@@ -0,0 +1,251 @@
...
+}
+
+/// Cache update handle.
+///
+/// Holds a cache write lock, the lock is released on drop.
+pub struct CacheUpdate<'a>(CacheUpdateInner<'a>);
+
+/// Internal state for [`CacheUpdate`].
+enum CacheUpdateInner<'a> {
+    Noop,
```
Comment (Dav1dde) [inline relay-cardinality/src/cache.rs:137]: Renamed from `Empty`

File: relay-cardinality/src/redis.rs [hunk 5]
```diff
@@ -1,167 +1,24 @@
...
-            self.entries.len(),
-            self.statuses.len(),
-            "expected same amount of entries as statuses"
-        );
-        std::iter::zip(self.entries, self.statuses)
-    }
-}
+// There are currently 5 metric namespaces and we know the cardinality scope is currently only the
+// metric namespace. Use this information to pre-allocate a few datastructures.
+const EXPECTED_SCOPE_COUNT: usize = 5;
```
Comment (Dav1dde) [inline relay-cardinality/src/redis.rs:21]: Factored into proper constant with documentation, no functional change

File: relay-cardinality/src/utils.rs
Comment (iker-barriocanal) [inline relay-cardinality/src/utils.rs:1]: `utils` -> `counter`? I don't really like "counter", but I don't find "utils" a perfect fit. Your call on this naming.

**Review by iker-barriocanal — APPROVED** [review]

File: relay-cardinality/src/utils.rs
Comment (Dav1dde) [reply relay-cardinality/src/utils.rs:1]: The reason I chose the `utils` module was that`HitCounter` is a very niche thing which is not related to the cardinality module directly (it just exists as a utility).

Maybe it fits in the `statsd` module?

Comment (iker-barriocanal) [reply relay-cardinality/src/utils.rs:1]: I don't fully like any alternative to be honest 😅 let's move forward with this and we can revisit it in the future if necessary.

**Pull request merged**